### PR TITLE
Switch macOS release to use self-hosted devices

### PR DESCRIPTION
**Pull request by antiagainst**

### CODE REVIEW

File: .github/workflows/build_package.yml [hunk 1]
```diff
@@ -52,12 +52,12 @@ jobs:
             build-package: py-tf-compiler-tools-pkg
             experimental: false
 
-          # Macos packages.
-          - runs-on: macos-11
+          # MacOS packages.
+          - runs-on: [self-hosted, os-family=macOS, runner-group=releaser]
```
Comment (ScottTodd) [inline .github/workflows/build_package.yml]: We might want to use a pattern like https://github.com/openxla/stablehlo/pull/1052 for self hosted runners

```
    runs-on: ${{ github.repository == 'openxla/iree' && '${SELF_HOSTED}' || '${PUBLIC}'  }}
```

**Review by ScottTodd — APPROVED** [review]

**Review by stellaraccident — APPROVED** [review]
lgtm modulo Scott's suggestion.

File: .github/workflows/build_package.yml [hunk 2]
```diff
@@ -140,6 +135,7 @@ jobs:
           package_suffix: ${{ github.event.inputs.package_suffix }}
           packages: "iree-runtime iree-runtime-instrumented"
           output_dir: "${{ github.workspace }}/bindist"
+          override_python_versions: "3.11"
```
Comment (GMNGeoffrey) [inline .github/workflows/build_package.yml:144]: Why does this change?

File: .github/workflows/build_package.yml (hunk 1 above)
Comment (GMNGeoffrey) [reply .github/workflows/build_package.yml]: Yeah I've been thinking about this. I would love to make it possible for a downstream to use their own self-hosted runners also. I think the best way to do this is to use repository variables: https://docs.github.com/en/actions/learn-github-actions/variables

**Review by GMNGeoffrey — COMMENTED** [review]
Looks like this is registered at the repo level. I'd like to get it registered at the org level where all our other runners are managed.

I also want to make sure that the maintenance process for this new runner is clear. Can you document how it was configured? Will it be auto-updating GitHub runner versions? What do we need to do to install new software? Who will take care of that? Who will debug when the build breaks? Can this build be dockerized?

**Comment by GMNGeoffrey** [comment]
I'm also a bit concerned with the integrity of the runner. We're producing release artifacts to be used by others and providing some guarantees that they are not malicious. This is using a single runner to perform releases that won't clear state between runs. Is software on this runner auto-updated by Apple (supply chain attack vector)? Does it get security updates? You'll note that our Linux release builds are *not* using our self-hosted runners. That's partially because I didn't feel sufficiently confident that we could ensure their security. I realize that slow mac releases are painful here, but I'd much prefer that we run this as a postsubmit bot rather than relying on it to produce any release artifacts. I think that would also provide more value in detecting breakages even if releases are slower and more stale.

**Comment by stellaraccident** [comment]
> I'm also a bit concerned with the integrity of the runner. We're producing release artifacts to be used by others and providing some guarantees that they are not malicious. This is using a single runner to perform releases that won't clear state between runs. Is software on this runner auto-updated by Apple (supply chain attack vector)? Does it get security updates? You'll note that our Linux release builds are _not_ using our self-hosted runners. That's partially because I didn't feel sufficiently confident that we could ensure their security. I realize that slow mac releases are painful here, but I'd much prefer that we run this as a postsubmit bot rather than relying on it to produce any release artifacts. I think that would also provide more value in detecting breakages even if releases are slower and more stale.

We have to have releases. If we don't do them automated, I'll have to build them on my personal machine and upload them.

The issue is not that Mac releases are slow but that they are presently impossible.

**Comment by GMNGeoffrey** [comment]
Ah Stella just explained that the release is literally timing out now... I had this mixed up with the `EXC_BAD_ACCESS` M1 bug. Ugh

File: .github/workflows/build_package.yml (hunk 1 above)
Comment (antiagainst) [reply .github/workflows/build_package.yml]: Ha, good point. Switched to do this!

File: .github/workflows/build_package.yml (hunk 2 above)
Comment (antiagainst) [reply .github/workflows/build_package.yml:144]: Suggestion from @powderluv. It makes sure we can build against the latest Python version, which is what we care right now. We can add more versions later if needed.

**Comment by antiagainst** [comment]
> Looks like this is registered at the repo level. I'd like to get it registered at the org level where all our other runners are managed.

Yeah I can migrate this to the org level.
 
> I also want to make sure that the maintenance process for this new runner is clear. Can you document how it was configured? Will it be auto-updating GitHub runner versions? What do we need to do to install new software? Who will take care of that? Who will debug when the build breaks? Can this build be dockerized?

Sure I can write up some docs detailing necessary information. I'll take the responsibility to manage these stuff, until it has been proven to not cause issues in an extended period of time, or we have dedicated staffing for it. :)

**Comment by antiagainst** [comment]
Manually triggered release with the latest change succeeded: https://github.com/openxla/iree/actions/runs/4308141110/jobs/7514056886

File: .github/workflows/build_package.yml (hunk 2 above)
Comment (GMNGeoffrey) [reply .github/workflows/build_package.yml:144]: Wait are we currently building with multiple python versions? Can this run on a single GitHub runner if we only build one?

Comment (stellaraccident) [reply .github/workflows/build_package.yml:144]: Each Python version adds ~a few minutes to the build. The majority is building the compiler, which is shared.

Comment (GMNGeoffrey) [reply .github/workflows/build_package.yml:144]: :hankey:

**Comment by antiagainst** [comment]
Moved runner to openxla org and installed pre/post job hooks. New runs succeeded: https://github.com/openxla/iree/actions/runs/4309691970. Landing this.

**Pull request merged**